File: midle/src/app/buy-option.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import {Configuration, Utils} from './app.constants';

/**
 * Buy option service
 */
@Injectable({
  providedIn: 'root'
})

/**
 * Buy option service
 * @author Diego Andre Poli <diegoandrepoli@gmail.com>
 */
export class BuyOptionService {

  /**
   * Server URL
   */
  private actionUrl: string;

  /**
   * Default constructor
   * @param http client injection
   */
  constructor(private http: HttpClient, private _utils: Utils, private _configuration: Configuration) {
    this.actionUrl = _configuration.server;
    let headers = this._utils.getHeaders();
  }

  /**
   * Get all buy options
   * @return list of buy option
   */
  public getAll<T>(): Observable<T> {
    return this.http.get<T>(this.actionUrl.concat('/buy-option'));
  }

  /**
   * Add buy option
   * @param buy option
   */
  public add<T>(id: string, item: Object ): Observable<T> {
    return this.http.post<T>( this.actionUrl.concat('/buy-option/add/deal/').concat(id), item);
  }

  /**
   * Update buy option
   * @param duy option
   */
  public update<T>( item: Object ): Observable<T> {
    return this.http.put<T>( this.actionUrl.concat('/buy-option/update'), item);
  }

  /**
   * Decrement quantity
   * @param id of buy option
   */
  public decrementQuantity<T>( id: string ): Observable<T> {
    return this.http.put<T>( this.actionUrl.concat('/buy-option/decrement/').concat(id), {});
  }

  /**
   * Delete buy option
   * @param id of buy option
   */
  public delete<T>( id: string ): Observable<T> {
    return this.http.delete<T>( this.actionUrl.concat('/buy-option/').concat(id));
  }

}

File: saler/src/main/java/com/saler/saler/buy/option/service/BuyOptionService.java
package com.saler.saler.buy.option.service;

import java.util.List;

import com.saler.saler.buy.option.entity.BuyOption;

/**
 * Deal service
 * @author Diego Andre Poli
 */
public interface BuyOptionService {

	/**
	 * Get deal by id
	 * @param id of deal
	 * @return deal
	 */
	BuyOption getById(Long id);
	
	/**
	 * Get all deals
	 * @return list of deals
	 */
	List<BuyOption> getAll();
	
	/**
	 * Add deal
	 * @param deal
	 * @return deal added
	 */
	BuyOption add(BuyOption deal);
	
	/**
	 * Remove deal
	 * @param id of deal
	 */
	void remove(Long id);
}
